File: test/apart_test_suite/c/calibrate2.c
#include <config.h>
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <sys/time.h>
#include "calibrate.h"
#include "work.h"

double
secs()
{
    struct timeval tv;
    gettimeofday( &tv, 0 );
    return tv.tv_sec + tv.tv_usec * 1.0e-6;
}

int
dcmp( const void* v1,
      const void* v2 )
{
    int i1 = *( double* )v1;
    int i2 = *( double* )v2;

    return i1 - i2;
}

#define MAX 13

int
main( int   argc,
      char* argv[] )
{
    double t1, t2, d[ MAX ], f;
    int    i, N;
    FILE*  out;

    printf( "calibration%d:\n", atoi( argv[ 1 ] ) );
    for ( i = 0; i < MAX; ++i )
    {
        printf( "." );
        fflush( stdout );
        t1 = secs();
        do_work( 3.42 );
        t2     = secs();
        d[ i ] = 3.42 / ( t2 - t1 );
    }
    printf( "\n" );

    qsort( d, MAX, sizeof( double ), dcmp );
    f = d[ MAX / 2 ];

    if ( fabs( f - 1.0 ) < 5.0e-4 )
    {
        printf( "%g wrong => close enough\n", f );
        return 1;
    }
    else
    {
        N = f * N_PER_SEC;
        printf( "%g wrong => using now %d\n\n", f, N );
    }

    if ( ( out = fopen( "calibrate.h", "w" ) ) == NULL )
    {
        perror( "calibrate.h" );
        return 1;
    }
    fprintf( out, "#define N_PER_SEC %d\n", N );
    fclose( out );

    return 0;
}

File: test/apart_test_suite/fortran/calibrate2.f90
PROGRAM CALIBRATE2

USE CALIBRATE
USE WORK

IMPLICIT NONE

INTEGER, PARAMETER :: MAX = 13

DOUBLE PRECISION :: t1, t2, f
INTEGER :: i, N, ios
DOUBLE PRECISION, DIMENSION(MAX) :: d

PRINT *, 'calibration:'
DO i = 1, MAX
  PRINT *, '.'
  t1 = secs()
  CALL do_work(3.42D0)
  t2 = secs()
  d(i) = 3.42D0 / (t2 - t1)
END DO

CALL sort_double(MAX, d)
f = d(MAX/2)
IF (ABS(f-1.0) .LT. 5.0E-4) THEN
  PRINT *, f, 'wrong => close enough'
  STOP
ELSE
  N = f * N_PER_SEC
  PRINT *, f, 'wrong => using now ', N
END IF

OPEN (UNIT=11, IOSTAT=ios, FILE='calibrate.f90', STATUS='REPLACE')
IF (ios .NE. 0) THEN
  PRINT *, 'error: failed to open the file calibrate.f90.'
  STOP
END IF
WRITE (11, '(A)') 'MODULE CALIBRATE'
WRITE (11, *) 'INTEGER, PARAMETER :: N_PER_SEC = ', N
WRITE (11, '(A)') 'END MODULE CALIBRATE'

CLOSE (UNIT=11)

CONTAINS
! ---------------------------------------------------------------------------

FUNCTION secs()

IMPLICIT NONE

DOUBLE PRECISION :: secs
INTEGER count, count_rate, count_max

CALL system_clock(count, count_rate, count_max)
secs = (count * 1D0) / count_rate

END FUNCTION secs

! ---------------------------------------------------------------------------

SUBROUTINE sort_double(n, array)

IMPLICIT NONE

INTEGER, INTENT(IN) :: n
DOUBLE PRECISION, DIMENSION(n), INTENT(IN OUT) :: array

LOGICAL :: no_exchange
INTEGER :: i
DOUBLE PRECISION :: temp

DO
  no_exchange = .TRUE.
  DO i = 1, n-1
    IF (array(i) > array(i+1)) THEN
      temp = array(i)
      array(i) = array(i+1)
      array(i+1) = temp
      no_exchange = .FALSE.
    END IF
  END DO
  IF (no_exchange) EXIT
END DO

END SUBROUTINE sort_double

! ---------------------------------------------------------------------------

END PROGRAM CALIBRATE2

